codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/smithypython/common/shapevisitor/conversionwriter/BaseConversionWriter.java

```diff
@@ -121,4 +129,34 @@ protected abstract void writeStructureShapeConverter(
   protected abstract void writeStringEnumShapeConverter(
     StringShape stringShapeWithEnumTrait
   );
+
+  /**
+   * Returns true if a conversion function should be written for the shape, false otherwise.
+   * Conversion functions are only written for "complex" shapes:
+   *  - StructureShapes ("complex" because StructureShapes can be recursive)
+   *    - except for non-AWS SDK StructureShapes with ErrorTrait; these aren't "complex"
```

Comment:
This won't be true forever as we should lift the smithy-dafny restriction that errors can only have a single string "message" member

Reply:
Sure -- I think that's a "cross that bridge when we come to it" situation